### ISSUE TITLE
Project not expandable anymore: Invalid read syntax: ". in wrong context"

### ISSUE THREAD

**Issue by xdl** [Alexander-Miller/treemacs#196]
I've run into a bug around projects 'breaking' in the treemacs pane. I can no longer expand it with `<TAB>`, which is giving an `Invalid read syntax: ". in wrong context"`.

Removing the project and re-adding it doesn't seem to work, nor does restarting Emacs. Any ideas on how to debug this further? I've tried disabling `treemacs-projectile` and `treemacs-evil` but getting the same result. I'm using the `20180511.327` version of treemacs.

It's happened to 2 projects so far but I haven't been able to figure out how to reproduce it yet, which I know isn't helpful :/

**Comment by Alexander-Miller** [comment]
Try enabling debug-on-error. I'll need a full stacktrace.

**Comment by Alexander-Miller** [comment]
Also what does the output of `(treemacs-current-workspace)` look like?

**Comment by xdl** [comment]
Stacktraces for the two unexpandable projects:

Project 1 (called `ge-evp`):
```
Debugger entered--Lisp error: (invalid-read-syntax ". in wrong context")
  read("((\"M\" . \"/Users/xiaodili/dev/ge-evp/docs/templateKeynote.org\")(\"M\" . \"/Users/xiaodili/dev/ge-evp/docs\")(\"M\" . \"/Users/xiaodili/dev/ge-evp/docs/templateTakehome.org\")(\"M\" . \"/Users/xiaodili/dev/ge-evp/notebooks/AnalysisWithRegion.ipynb\")(\"M\" . \"/Users/xiaodili/dev/ge-evp/notebooks\")(\"M\" . \"/Users/xiaodili/dev/ge-evp/\"notebooks/EVP Themes.ipynb\"\")(\"M\" . \"/Users/xiaodili/dev/ge-evp/\"notebooks\")(\"??\" . \"/Users/xiaodili/dev/ge-evp/data/Percentage splits.xlsx\")(\"M\" . \"/Users/xiaodili/dev/ge-evp/data\")(\"??\" . \"/Users/xiaodili/dev/ge-evp/mockup\")(\"?\" . \"/Users/xiaodili/dev/ge-evp/mockup/index.html\")(\"?\" . \"/Users/xiaodili/dev/ge-evp/mockup/assets\")(\"?\" . \"/Users/xiaodili/dev/ge-evp/mockup/assets/style.css\")(\"??\" . \"/Users/xiaodili/dev/ge-evp/notebooks/3reasons_by_region.html\")(\"??\" . \"/Users/xiaodili/dev/ge-evp/notebooks/FreetextExploration.ipynb\")(\"??\" . \"/Users/xiaodili/dev/ge-evp/notebooks/notebook.tex\")(\"??\" . \"/Users/xiaodili/dev/ge-evp/notebooks/output_34_0.png\")(\"??\" . \"/Users/xiaodili/dev/ge-evp/notebooks/output_35_0.png\")(\"??\" . \"/Users/xiaodili/dev/ge-evp/notebooks/testContainer.html\")(\"??\" . \"/Users/xiaodili/dev/ge-evp/notebooks/testplot.html\")(\"??\" . \"/Users/xiaodili/dev/ge-evp/webapp\")(\"?\" . \"/Users/xiaodili/dev/ge-evp/webapp/typings.d.ts\")(\"?\" . \"/Users/xiaodili/dev/ge-evp/webapp/webpack.dev.js\")(\"?\" . \"/Users/xiaodili/dev/ge-evp/webapp/tslint.json\")(\"?\" . \"/Users/xiaodili/dev/ge-evp/webapp/webpack.common.js\")(\"!!\" . \"/Users/xiaodili/dev/ge-evp/.DS_Store\")(\"!!\" . \"/Users/xiaodili/dev/ge-evp/data/.DS_Store\")(\"!!\" . \"/Users/xiaodili/dev/ge-evp/notebooks/.ipynb_checkpoints\")(\"!!\" . \"/Users/xiaodili/dev/ge-evp/venv\"))")
  treemacs--git-status-parse-function(#<process Process Future>)
  treemacs--create-branch("/Users/xiaodili/dev/ge-evp" 1 #<process Process Future> nil #<marker (moves after insertion) at 26 in  *Treemacs-Framebuffer-1*>)
  treemacs--expand-root-node(#<marker (moves after insertion) at 26 in  *Treemacs-Framebuffer-1*>)
  treemacs-toggle-node(nil)
  treemacs-TAB-action(nil)
  funcall-interactively(treemacs-TAB-action nil)
  call-interactively(treemacs-TAB-action nil nil)
  command-execute(treemacs-TAB-action)Debugger entered--Lisp error: (invalid-read-syntax ". in wrong context")
  read("((\"M\" . \"/Users/xiaodili/dev/ge-evp/docs/templateKeynote.org\")(\"M\" . \"/Users/xiaodili/dev/ge-evp/docs\")(\"M\" . \"/Users/xiaodili/dev/ge-evp/docs/templateTakehome.org\")(\"M\" . \"/Users/xiaodili/dev/ge-evp/notebooks/AnalysisWithRegion.ipynb\")(\"M\" . \"/Users/xiaodili/dev/ge-evp/notebooks\")(\"M\" . \"/Users/xiaodili/dev/ge-evp/\"notebooks/EVP Themes.ipynb\"\")(\"M\" . \"/Users/xiaodili/dev/ge-evp/\"notebooks\")(\"??\" . \"/Users/xiaodili/dev/ge-evp/data/Percentage splits.xlsx\")(\"M\" . \"/Users/xiaodili/dev/ge-evp/data\")(\"??\" . \"/Users/xiaodili/dev/ge-evp/mockup\")(\"?\" . \"/Users/xiaodili/dev/ge-evp/mockup/index.html\")(\"?\" . \"/Users/xiaodili/dev/ge-evp/mockup/assets\")(\"?\" . \"/Users/xiaodili/dev/ge-evp/mockup/assets/style.css\")(\"??\" . \"/Users/xiaodili/dev/ge-evp/notebooks/3reasons_by_region.html\")(\"??\" . \"/Users/xiaodili/dev/ge-evp/notebooks/FreetextExploration.ipynb\")(\"??\" . \"/Users/xiaodili/dev/ge-evp/notebooks/notebook.tex\")(\"??\" . \"/Users/xiaodili/dev/ge-evp/notebooks/output_34_0.png\")(\"??\" . \"/Users/xiaodili/dev/ge-evp/notebooks/output_35_0.png\")(\"??\" . \"/Users/xiaodili/dev/ge-evp/notebooks/testContainer.html\")(\"??\" . \"/Users/xiaodili/dev/ge-evp/notebooks/testplot.html\")(\"??\" . \"/Users/xiaodili/dev/ge-evp/webapp\")(\"?\" . \"/Users/xiaodili/dev/ge-evp/webapp/typings.d.ts\")(\"?\" . \"/Users/xiaodili/dev/ge-evp/webapp/webpack.dev.js\")(\"?\" . \"/Users/xiaodili/dev/ge-evp/webapp/tslint.json\")(\"?\" . \"/Users/xiaodili/dev/ge-evp/webapp/webpack.common.js\")(\"!!\" . \"/Users/xiaodili/dev/ge-evp/.DS_Store\")(\"!!\" . \"/Users/xiaodili/dev/ge-evp/data/.DS_Store\")(\"!!\" . \"/Users/xiaodili/dev/ge-evp/notebooks/.ipynb_checkpoints\")(\"!!\" . \"/Users/xiaodili/dev/ge-evp/venv\"))")
  treemacs--git-status-parse-function(#<process Process Future>)
  treemacs--create-branch("/Users/xiaodili/dev/ge-evp" 1 #<process Process Future> nil #<marker (moves after insertion) at 26 in  *Treemacs-Framebuffer-1*>)
  treemacs--expand-root-node(#<marker (moves after insertion) at 26 in  *Treemacs-Framebuffer-1*>)
  treemacs-toggle-node(nil)
  treemacs-TAB-action(nil)
  funcall-interactively(treemacs-TAB-action nil)
  call-interactively(treemacs-TAB-action nil nil)
  command-execute(treemacs-TAB-action)
```

Project 2 (`dev-talk-data-science`):
```
Debugger entered--Lisp error: (invalid-read-syntax ". in wrong context")
  read("((\"D\" . \"/Users/xiaodili/dev/dev-talk-data-science/\"notebooks/Data Science in Jupyter Whirlwind Tour.ipynb\"\")(\"M\" . \"/Users/xiaodili/dev/dev-talk-data-science/\"notebooks\")(\"??\" . \"/Users/xiaodili/dev/dev-talk-data-science/notebooks/Whirlwind introduction to Data Science.ipynb\")(\"M\" . \"/Users/xiaodili/dev/dev-talk-data-science/notebooks\")(\"!!\" . \"/Users/xiaodili/dev/dev-talk-data-science/notebooks/.ipynb_checkpoints\")(\"!!\" . \"/Users/xiaodili/dev/dev-talk-data-science/venv\"))")
  treemacs--git-status-parse-function(#<process Process Future>)
  treemacs--create-branch("/Users/xiaodili/dev/dev-talk-data-science" 1 #<process Process Future> nil #<marker (moves after insertion) at 75 in  *Treemacs-Framebuffer-1*>)
  treemacs--expand-root-node(#<marker (moves after insertion) at 75 in  *Treemacs-Framebuffer-1*>)
  treemacs-toggle-node(nil)
  treemacs-TAB-action(nil)
  funcall-interactively(treemacs-TAB-action nil)
  call-interactively(treemacs-TAB-action nil nil)
  command-execute(treemacs-TAB-action)
```

Evaluating `(treemacs-current-workspace)`:

```
[cl-struct-treemacs-workspace "Default Workspace" ([cl-struct-treemacs-project "dotfiles" "/Users/xiaodili/dotfiles"] [cl-struct-treemacs-project "devlogs" "/Users/xiaodili/Documents/devlogs"] [cl-struct-treemacs-project "ge-evp" "/Users/xiaodili/dev/ge-evp"] [cl-struct-treemacs-project "knowledgebase-wiki" "/Users/xiaodili/dev/knowledgebase-wiki"] [cl-struct-treemacs-project "orgs" "/Users/xiaodili/Documents/orgs"] [cl-struct-treemacs-project "notes" "/Users/xiaodili/Documents/notes"] [cl-struct-treemacs-project "dev-talk-data-science!" "/Users/xiaodili/dev/dev-talk-data-science"] [cl-struct-treemacs-project "dtds" "/Users/xiaodili/dev/dtds"] [cl-struct-treemacs-project "smartlight" "/Users/xiaodili/dev/smartlight"])]
```

**Comment by Alexander-Miller** [comment]
> (\"M\" . \"/Users/xiaodili/dev/ge-evp/\"notebooks\")

And here's the problem. The python script that determines your files' git status outputs a string that is directly read by the elisp interpreter. This process is tripped up by quotes being part of the path.

Unfortunately I cannot correct this either. I tried a bunch of quote-escaping combinations in the script, but the issue always persists. I guess you've 2 options here: don't use quotes in filenames or don't use extended git mode (simple works fine).

**Comment by xdl** [comment]
Yup setting it to simple git mode has fixed it. If I have a bit of time next weekend I'll try to have a poke around the script but this helps a lot, thanks!

**Comment by Alexander-Miller** [comment]
You don't necessarily need to look at the script. The short version is that you need to get this expression working, such that you read a string that correctly contains quotes: 
`(read "((\"M\" . \"/Users/xiaodili/dev/ge-evp/\"notebooks\"))")`

**Comment by Compro-Prasad** [comment]
I think doing a bit of preprocessing in both the script and the elisp would do this if escaping it isn't helping.

For example each `"` will be converted to `# ` [1 hash and 1 space]. ` ` [1 space] and `#` can be converted to `  ` [2 spaces] and `##`.

**Comment by Compro-Prasad** [comment]
Can you reference me the lines in elisp where this script is referenced?

**Comment by Alexander-Miller** [comment]
I don't like the performance implications of that at all. The new deferred git-mode has taken the parsing somewhat outside the most critical path, but it's still a lot of work for what I see as a very out of the way edge case.

If you want to try it all the relevant code is in treemacs-async.el. Due to the simple/extended git-mode choice there a bit of `fset` magic going on: https://github.com/Alexander-Miller/treemacs/blob/master/src/elisp/treemacs-async.el#L196. 
Basically your entry point looks like this:

```
(-> "~/Documents/git/treemacs"
    (treemacs--git-status-process-extended)
    (treemacs--parse-git-status-extended))
```

**Comment by Alexander-Miller** [comment]
Closing due to inactivity.
